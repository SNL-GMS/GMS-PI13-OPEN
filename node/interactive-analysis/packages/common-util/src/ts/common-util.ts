export { KeyValue } from './types/key-value';
export * from './util/compose-util';
export * from './util/delay-execution-util';
export * from './util/environment-util';
export * from './util/file-system-util';
export * from './util/json-util';
export * from './util/object-to-graphql-util';
export * from './util/random-number-util';
export * from './util/string-util';
export * from './util/time-util';
export * from './util/uuid-util';
export * from './util/formatting-util';
export * from './util/sleep-util';
export * from './util/window-util';
