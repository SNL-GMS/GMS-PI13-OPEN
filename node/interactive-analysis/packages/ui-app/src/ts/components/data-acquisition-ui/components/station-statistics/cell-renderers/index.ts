import {
  ChannelEnvironmentCellRenderer,
  StationEnvironmentCellRenderer
} from './environment-cell-renderer';
import { ChannelLagCellRenderer, StationLagCellRenderer } from './lag-cell-renderer';
import { ChannelMissingCellRenderer, StationMissingCellRenderer } from './missing-cell-renderer';
import { StationNameCellRenderer } from './station-cell-renderer';
import {
  ChannelTimelinessCellRenderer,
  StationTimelinessCellRenderer
} from './timeliness-cell-renderer';
export {
  ChannelTimelinessCellRenderer,
  StationTimelinessCellRenderer,
  ChannelEnvironmentCellRenderer,
  StationEnvironmentCellRenderer,
  ChannelLagCellRenderer,
  StationLagCellRenderer,
  ChannelMissingCellRenderer,
  StationMissingCellRenderer,
  StationNameCellRenderer
};
