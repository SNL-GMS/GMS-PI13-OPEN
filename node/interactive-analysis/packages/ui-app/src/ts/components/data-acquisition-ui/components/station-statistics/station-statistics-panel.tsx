import { IconNames } from '@blueprintjs/icons';
import { SohTypes } from '@gms/common-graphql';
import {
  isEnvironmentalIssue,
  StationAggregateType,
  UiStationSoh
} from '@gms/common-graphql/lib/graphql/soh/types';
import { CenterIcon, HorizontalDivider, ToolbarTypes } from '@gms/ui-core-components';
import { Cancelable, cloneDeep, debounce, flatMap, isEqual, max, orderBy } from 'lodash';
import memoizeOne from 'memoize-one';
import * as React from 'react';
import { dataAcquisitionUserPreferences } from '~components/data-acquisition-ui/config';
import { messageConfig } from '~components/data-acquisition-ui/config/message-config';
import { StationDeselectHandler } from '~components/data-acquisition-ui/shared/table/station-deselect-handler';
import {
  acknowledgeContextMenu,
  getWorseStatus,
  isSohStationStaleTimeMS,
  sharedSohTableClasses
} from '~components/data-acquisition-ui/shared/table/utils';
import {
  getLatestSohTime,
  initialFiltersToDisplay,
  SohToolbar
} from '~components/data-acquisition-ui/shared/toolbars/soh-toolbar';
import { DropZone } from '../../shared/table/drop-zone';
import { Columns } from './column-definitions';
import {
  StationStatisticsContext,
  StationStatisticsContextData
} from './station-statistics-context';
import { StationStatisticsTable } from './station-statistics-table';
import {
  StationStatisticsPanelProps,
  StationStatisticsPanelState,
  StationStatisticsRow,
  StationStatisticsRowClickedEvent
} from './types';

/**
 * Renders the toolbar, divider, and station statistics display.
 */
export class StationStatisticsPanel extends React.Component<
  StationStatisticsPanelProps,
  StationStatisticsPanelState
> {
  /** The context type */
  public static contextType: React.Context<StationStatisticsContextData> = StationStatisticsContext;

  /** The context wrapped around the panel, used for mutation and access to the GL Container */
  public context: React.ContextType<typeof StationStatisticsContext>;

  /**
   * A memoized function for generating the table rows.
   * @param stationSohs station soh data
   * @param groupSelected the selected group
   * @returns an array of station statistics rows
   */
  private readonly memoizedGenerateTableRows: (
    stationSohs: SohTypes.UiStationSoh[],
    groupSelected: string
  ) => StationStatisticsRow[];

  /** Default group selector text for the dropdown filter */
  private readonly defaultGroupSelectorText: string = 'All Groups';

  /** A ref to the bottom that toggles checkbox dropdown in the toolbar */
  private highlightButtonRef: HTMLDivElement = null;

  /**
   * How frequently to attempt to set the highlight
   */
  private readonly setStateTryMs: number = 200;

  /**
   * Set the highlight state on a debounced function so that many
   * event calls do not clog up the event stack.
   */
  private readonly debouncedSetStateHighlight: (() => void) & Cancelable = debounce(
    () => {
      this.setState({ isHighlighted: true });
    },
    this.setStateTryMs,
    { leading: true }
  );

  /** The reference to the acknowledged table */
  private tableAcknowledged: StationStatisticsTable;

  /** The reference to the unacknowledged table */
  private tableUnacknowledged: StationStatisticsTable;

  /**
   * constructor
   */
  public constructor(props: StationStatisticsPanelProps) {
    super(props);
    this.memoizedGenerateTableRows =
      typeof memoizeOne === 'function'
        ? memoizeOne(
            this.generateTableData,
            /* tell memoize to use a deep comparison for complex objects */
            isEqual
          )
        : this.generateTableData;
    this.state = {
      isHighlighted: false,
      statusesToDisplay: initialFiltersToDisplay,
      groupSelected: undefined,
      columnsToDisplay: new Map(
        Object.values(Columns)
          // all columns are visible by default
          .map(v => [v, true])
      )
    };
  }

  // ***************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ***************************************

  /**
   * Renders the component.
   */
  public render() {
    /** Default top row height in pixels */
    const defaultTopRowHeightPx = 425;
    /** Icon size */
    const iconSize = 50;
    /** The selected group to filter on */
    const groupSelector = this.generateGroupSelector(
      this.props.stationGroups,
      this.state.groupSelected
    );

    const columnsToDisplayCheckBoxDropdown = this.generateColumnDisplaySelector();

    /** Will filter if a group is selected otherwise use the props passed in */
    const dataForBothTables = this.state.groupSelected
      ? this.memoizedGenerateTableRows(
          this.props.stationSohs,
          this.state.groupSelected
        ).filter(data =>
          data.stationGroups.find(group => group.groupName === this.state.groupSelected)
        )
      : this.memoizedGenerateTableRows(this.props.stationSohs, this.state.groupSelected);

    /** Filters data based on needing acknowledgement, this data is used for the top table */
    const needsAttentionSohData = dataForBothTables.filter(station => station.needsAttention);

    /** Filters the soh data that does not need attention, this data is used for the bottom table */
    const doesNotNeedAttentionSohData = dataForBothTables
      .filter(station => !station.needsAttention)
      .filter(row => this.state.statusesToDisplay.get(row.stationData.stationCapabilityStatus));

    return (
      <React.Fragment>
        <div className="soh-toolbar-container">
          <SohToolbar
            statusesToDisplay={this.state.statusesToDisplay}
            setStatusesToDisplay={this.setStatusesToDisplay}
            leftItems={[groupSelector, columnsToDisplayCheckBoxDropdown]}
            rightItems={[]}
            statusFilterText={messageConfig.labels.sohToolbar.filterStatuses}
            statusFilterTooltip={messageConfig.tooltipMessages.sohToolbar.selectStatuses}
            updatedAt={getLatestSohTime(this.props.stationSohs)}
            updateIntervalSecs={this.props.updateIntervalSecs}
            toggleHighlight={this.toggleHighlight}
            sohStationStaleTimeMS={this.context.sohStationStaleTimeMS}
            displayTimeWarning={isSohStationStaleTimeMS(
              getLatestSohTime(this.props.stationSohs),
              this.context.sohStationStaleTimeMS
            )}
          />
        </div>
        <HorizontalDivider
          topHeightPx={defaultTopRowHeightPx}
          sizeRange={dataAcquisitionUserPreferences.stationStatisticsMinContainerRange}
          top={
            <StationDeselectHandler
              setSelectedStationIds={(ids: string[]) => this.context.setSelectedStationIds(ids)}
              className={`station-statistics-table-container station-statistics-table-container--needs-attention ${sharedSohTableClasses}`}
              dataCy="soh-unacknowledged"
            >
              <div className="soh-table-label">Needs Attention</div>
              {needsAttentionSohData.length !== 0 ? (
                <StationStatisticsTable
                  ref={ref => {
                    this.tableUnacknowledged = ref;
                  }}
                  id="soh-unacknowledged"
                  tableData={needsAttentionSohData}
                  onRowClicked={this.onRowClicked}
                  acknowledgeContextMenu={this.acknowledgeContextMenu}
                />
              ) : (
                <CenterIcon
                  iconName={IconNames.TICK_CIRCLE}
                  description={'Nothing to acknowledge'}
                  iconSize={iconSize}
                  className={'table-background-icon'}
                />
              )}
            </StationDeselectHandler>
          }
          bottom={
            <StationDeselectHandler
              setSelectedStationIds={(ids: string[]) => this.context.setSelectedStationIds(ids)}
              className={`station-statistics-table-container station-statistics-table-container--acknowledged gms-drop-zone
                  ${sharedSohTableClasses}
                  ${
                    this.state.isHighlighted
                      ? 'station-statistics-table-container--highlighted'
                      : ''
                  }`}
              dataCy="soh-acknowledged"
            >
              <DropZone onDrop={(payload: string[]) => this.cellDrop(payload, this.context)}>
                <StationStatisticsTable
                  ref={ref => {
                    this.tableAcknowledged = ref;
                  }}
                  id="soh-acknowledged"
                  tableData={doesNotNeedAttentionSohData}
                  onRowClicked={this.onRowClicked}
                  acknowledgeContextMenu={(stationNames: string[]) =>
                    this.acknowledgeContextMenu(stationNames)
                  }
                  suppressContextMenu={true}
                  highlightDropZone={this.debouncedSetStateHighlight}
                />
              </DropZone>
            </StationDeselectHandler>
          }
        />
      </React.Fragment>
    );
  }

  // ***************************************
  // END REACT COMPONENT LIFECYCLE METHODS
  // ***************************************

  /**
   * Handles row selection and updated the redux state with station ID for sync selection
   * @param event StationStatisticsRowClickedEvent
   */
  private readonly onRowClicked = (event: StationStatisticsRowClickedEvent): void => {
    const stationNames: string[] = event.api.getSelectedRows().map(row => row.id);

    this.context.setSelectedStationIds(stationNames);
    if (this.tableUnacknowledged) {
      this.tableUnacknowledged.updateRowSelection(stationNames);
    }

    if (this.tableAcknowledged) {
      this.tableAcknowledged.updateRowSelection(stationNames);
    }
  }

  private readonly acknowledgeContextMenu = (stationNames: string[]): JSX.Element =>
    acknowledgeContextMenu(
      stationNames,
      this.props.stationSohs,
      this.context.sohStationStaleTimeMS,
      (names: string[], comment?: string) => this.context.acknowledgeSohStatus(names, comment)
    )

  /**
   * Cell drag logic, gets data from transfer object and calls acknowledgeSohStatus
   * @param event React.DragEvent<HTMLDivElement>
   * @param context context data for an soh panel
   */
  private readonly cellDrop = (
    stationNames: string[],
    context: StationStatisticsContextData
  ): void => {
    this.debouncedSetStateHighlight.cancel();
    this.setState({ isHighlighted: false });
    context.acknowledgeSohStatus(stationNames);
    context.setSelectedStationIds(stationNames);
  }

  /**
   * Generates the drop down for the groups
   * @param groups station groups list
   * @param groupSelected current group selected
   * @returns a toolbar item of type ToolbarTypes.ToolbarItem
   */
  private readonly generateGroupSelector = (
    groups: SohTypes.StationGroupSohStatus[],
    groupSelected: string
  ): ToolbarTypes.ToolbarItem => {
    const groupOptions = { default: this.defaultGroupSelectorText };
    orderBy(groups, 'priority', 'asc').forEach(group => {
      groupOptions[group.id] = group.stationGroupName;
    });

    return {
      rank: 2,
      label: 'Select a Group',
      tooltip: 'Show only a single group',
      type: ToolbarTypes.ToolbarItemType.Dropdown,
      onChange: e => {
        this.setGroupSelected(e !== this.defaultGroupSelectorText ? e : undefined);
      },
      dropdownOptions: groupOptions,
      value: groupSelected ? groupSelected : this.defaultGroupSelectorText,
      cyData: 'station-statistics-group-selector',
      widthPx: 150
    };
  }

  /**
   * Generates the drop down for selecting which columns are visible.
   * @returns a toolbar item of type ToolbarTypes.ToolbarItem
   */
  private readonly generateColumnDisplaySelector = (): ToolbarTypes.ToolbarItem => {
    const columnsToDisplayCheckBoxDropdown: ToolbarTypes.CheckboxDropdownItem = {
      enumOfKeys: Object.values(Columns)
        // do not allow the station column to be hidden
        .filter(v => v !== Columns.Station),
      label: 'Show columns',
      menuLabel: 'Show columns',
      rank: 3,
      widthPx: 150,
      type: ToolbarTypes.ToolbarItemType.CheckboxList,
      tooltip: 'Set which columns are visible',
      values: this.state.columnsToDisplay,
      // tslint:disable-next-line: no-console
      onChange: this.setColumnsToDisplay,
      cyData: 'filter-column'
    };
    return columnsToDisplayCheckBoxDropdown;
  }

  /**
   * @param stationSoh the soh data for the station belonging to a row
   * @param monitorType either a monitor type, or 'ENVIRONMENT' in the case of the environment cell.
   * Note that 'ENVIRONMENT' string is used instead of an enum because there is no single enum corresponding
   * to the channel issues cell. Instead, that cell is a rollup of all environmental issues.
   */
  private readonly generateChannelCellData = (
    stationSoh: SohTypes.UiStationSoh,
    monitorType: SohTypes.SohMonitorType | 'ENVIRONMENT'
  ) => {
    const contributorData =
      monitorType === 'ENVIRONMENT'
        ? this.getEnvRollup(stationSoh)
        : this.getContributorForSohMonitorType(stationSoh, monitorType);
    return {
      value: contributorData && contributorData.valuePresent ? contributorData.value : undefined,
      status: contributorData ? contributorData.statusSummary : undefined,
      isContributing: contributorData ? contributorData.contributing : false
    };
  }

  /**
   * Generates table data
   * @param props the props
   * @returns an array of table data rows
   */
  private readonly generateTableData = (
    stationSohs: SohTypes.UiStationSoh[],
    groupSelected: string
  ): StationStatisticsRow[] =>
    cloneDeep(
      flatMap(stationSohs, stationSoh => {
        const stationData = {
          stationName: stationSoh?.stationName,
          stationStatus: stationSoh?.sohStatusSummary,
          stationCapabilityStatus: groupSelected
            ? stationSoh.stationGroups.find(group => group.groupName === groupSelected)
                ?.sohStationCapability
            : this.getWorstStationCapabilityStatus(stationSoh)
        };

        // Data for channel detail

        // Channel level detail/statistics
        const channelLag = this.generateChannelCellData(stationSoh, SohTypes.SohMonitorType.LAG);
        const channelMissing = this.generateChannelCellData(
          stationSoh,
          SohTypes.SohMonitorType.MISSING
        );
        const channelEnvironment = this.generateChannelCellData(stationSoh, 'ENVIRONMENT');
        const channelTimeliness = this.generateChannelCellData(
          stationSoh,
          SohTypes.SohMonitorType.TIMELINESS
        );

        // Station level details/statistics
        // Average transmission time for acquired data samples.
        const stationLag = this.getStationStats(stationSoh, StationAggregateType.LAG);

        // Gets the total percentage of 'bad' indicators across all channels and all environmental monitors.
        const stationEnvironment = this.getStationStats(
          stationSoh,
          StationAggregateType.ENVIRONMENTAL_ISSUES
        );

        // Returns the total percentage of missing data across all channels.
        const stationMissing = this.getStationStats(stationSoh, StationAggregateType.MISSING);

        // Latest data sample time that has been acquired on any channel.
        const stationTimeliness = this.getStationStats(stationSoh, StationAggregateType.TIMELINESS);

        return {
          id: `${stationSoh.stationName}`,
          stationData,
          stationGroups: groupSelected
            ? [stationSoh.stationGroups.find(group => group.groupName === groupSelected)].filter(
                group => group !== undefined
              )
            : stationSoh.stationGroups,
          needsAcknowledgement: stationSoh.needsAcknowledgement,
          needsAttention: stationSoh.needsAttention,
          channelLag,
          channelMissing,
          channelEnvironment,
          channelTimeliness,
          stationLag,
          stationMissing,
          stationEnvironment,
          stationTimeliness
        };
      }).sort((stationA, stationB) => stationA.id.localeCompare(stationB.id))
    )

  /**
   * Returns the contributor for soh monitor type by station status
   * @param stationStatus SohTypes.UiStationSoh
   * @param monitorType: SohTypes.SohMonitorType
   * @returns SohTypes.SohContributor
   */
  private readonly getContributorForSohMonitorType = (
    stationStatus: SohTypes.UiStationSoh,
    monitorType: SohTypes.SohMonitorType
  ): SohTypes.SohContributor =>
    stationStatus.statusContributors.find(contributor => contributor.type === monitorType)

  /**
   * Toggles the highlight for the button that toggles the dropdown filter list
   * Calls a method that updates the toggleHighlight state
   * @param ref ref to the filter button
   */
  private readonly toggleHighlight = (ref: HTMLDivElement): void => {
    this.setIsHighlighted(!this.state.isHighlighted);
    this.highlightButtonRef = ref ?? this.highlightButtonRef;
    this.highlightButtonRef?.classList?.toggle('isHighlighted');
  }

  /**
   * Updates the state for isHighlighted
   * @param isHighlightedValue value to update the state with
   */
  private readonly setIsHighlighted = (isHighlightedValue: boolean): void => {
    this.setState({
      isHighlighted: isHighlightedValue
    });
  }

  /**
   * Updates the state for groupSelected
   * @param groupSelected value to update the state with
   */
  private readonly setGroupSelected = (groupSelected: string): void => {
    this.setState({
      groupSelected
    });
  }

  /**
   * Updates the state for statusesToDisplay
   * @param statusesToDisplay value to update the state with
   */
  private readonly setStatusesToDisplay = (statusesToDisplay: Map<any, boolean>): void => {
    this.setState({
      statusesToDisplay
    });
  }

  /**
   * Updates the state for which columns are displayed
   * @param columnsToDisplay value to update the state with
   */
  private readonly setColumnsToDisplay = (columnsToDisplay: Map<Columns, boolean>): void => {
    this.setState({ columnsToDisplay });
    this.tableAcknowledged?.updateColumnVisibility(columnsToDisplay);
    this.tableUnacknowledged?.updateColumnVisibility(columnsToDisplay);
  }

  /**
   * Gets the rolled up data for the environment column
   * @param stationStatus station soh to get environment rollup for
   */
  private readonly getEnvRollup = (
    stationStatus: SohTypes.UiStationSoh
  ): SohTypes.SohContributor => {
    const envContributors = stationStatus.statusContributors.filter(contributor =>
      isEnvironmentalIssue(contributor.type)
    );
    const worseStatus = envContributors
      .map(envContributor => envContributor.statusSummary)
      .reduce(getWorseStatus, SohTypes.SohStatusSummary.NONE);
    const worseValue = max(
      envContributors
        .filter(contributor => contributor.statusSummary === worseStatus)
        .map(contributor => (contributor.valuePresent ? contributor.value : undefined))
    );

    // return rollup of worse status - and the worst value of the worst status - type is not used
    return {
      contributing: envContributors.filter(c => c.contributing).length > 0,
      statusSummary: worseStatus,
      type: undefined,
      value: worseValue,
      valuePresent: worseValue !== undefined
    };
  }

  /**
   * Gets the worst capability status for the given station
   * @param station station soh
   */
  private readonly getWorstStationCapabilityStatus = (station: UiStationSoh) =>
    station.stationGroups
      .map(group => group.sohStationCapability)
      .reduce(getWorseStatus, SohTypes.SohStatusSummary.NONE)

  /**
   * Finds any one of envIssues, lag, missing, or timeliness stat for a given station
   * @param stationStatus - ui station soh we will find stats for
   * @param stationAggType - the stat we wish to find
   * @returns the percentage or seconds of the stat
   */
  private readonly getStationStats = (
    stationStatus: SohTypes.UiStationSoh,
    stationAggType: StationAggregateType
  ): number => {
    const stationAggregate = stationStatus.allStationAggregates.filter(
      stAgg => stAgg.aggregateType === stationAggType
    )[0];
    return stationAggregate && stationAggregate.valuePresent ? stationAggregate.value : undefined;
  }
}
