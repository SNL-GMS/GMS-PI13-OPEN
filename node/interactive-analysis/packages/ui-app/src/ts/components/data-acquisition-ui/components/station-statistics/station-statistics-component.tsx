import { SohTypes } from '@gms/common-graphql';
import { MILLISECONDS_IN_SECOND, millisToStringWithMaxPrecision } from '@gms/common-util';
import { UILogger } from '@gms/ui-apollo';
import { Toaster } from '@gms/ui-util';
import isEqual from 'lodash/isEqual';
import React from 'react';
import { BaseDisplay } from '~components/common-ui/components/base-display';
import { isAcknowledgeEnabled } from '~components/data-acquisition-ui/shared/table/utils';
import { StationStatisticsContext } from './station-statistics-context';
import { StationStatisticsPanel } from './station-statistics-panel';
import { StationStatisticsProps } from './types';

/**
 * Connected component that provides context and props to SOH Panel
 */
export class StationStatisticsComponent extends React.Component<StationStatisticsProps, {}> {
  /*
   * Series of member variables to report lag statistics every 30 mins
   */

  /** Represents a half hour (30 minutes) in milliseconds */
  private readonly MS_HALF_HOUR: number = 1800000;

  /** the soh soh lag report period; how frequently the to report */
  private readonly sohLagReportPeriod: number = this.MS_HALF_HOUR;

  /** the last time the soh report was sent (reported) */
  private lastSohReport: number = Date.now();

  /** the min soh lag value */
  private minSohLag: number;

  /** the max soh lag value */
  private maxSohLag: number;

  /** the average soh lag value */
  private avgSohLag: number = 0;

  /** the soh station count - used to calculate the average */
  private sohStationsCount: number = 0;

  // Map to not check StationSOH already sent from API Gateway
  private readonly sohCheckDisplayCriteriaMap: Map<string, number>;

  /**
   * constructor
   */
  public constructor(props) {
    super(props);
    this.sohCheckDisplayCriteriaMap = new Map<string, number>();
  }

  // ***************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ***************************************

  /**
   * React lifecycle `componentDidUpdate`.
   * Called immediately after updating occurs. Not called for the initial render.
   *
   * @param prevProps the previous props
   * @param prevState the previous state
   */
  public componentDidUpdate(prevProps: StationStatisticsProps, prevState: StationStatisticsProps) {
    if (!isEqual(prevProps.sohStatus, this.props)) {
      // Check SOH station entries to see if the lag exceeds the time limit
      this.checkSOHDisplayCriteria(this.props.sohStatus.stationAndStationGroupSoh.stationSoh);
    }
  }

  /**
   * React lifecycle `render`.
   * Renders the component.
   */
  public render() {
    return (
      <StationStatisticsContext.Provider
        value={{
          acknowledgeSohStatus: this.acknowledgeSohStatus,
          quietTimerMs: this.props.uiConfigurationQuery.uiAnalystConfiguration
            .acknowledgementQuietDuration,
          sohStationStaleTimeMS: this.props.uiConfigurationQuery.uiAnalystConfiguration
            .sohStationStaleTimeMS,
          updateIntervalSecs: this.props.uiConfigurationQuery.uiAnalystConfiguration
            .reprocessingPeriod,
          selectedStationIds: this.props.selectedStationIds,
          setSelectedStationIds: this.props.setSelectedStationIds
        }}
      >
        <BaseDisplay
          glContainer={this.props.glContainer}
          className="soh-divider-container station-statistics-display drop-zone__wrapper"
        >
          <StationStatisticsPanel
            stationGroups={this.props.sohStatus.stationAndStationGroupSoh.stationGroups}
            stationSohs={this.props.sohStatus.stationAndStationGroupSoh.stationSoh}
            updateIntervalSecs={
              this.props.uiConfigurationQuery.uiAnalystConfiguration.reprocessingPeriod
            }
            selectedStationIds={this.props.selectedStationIds}
            setSelectedStationIds={ids => this.props.setSelectedStationIds(ids)}
          />
        </BaseDisplay>
      </StationStatisticsContext.Provider>
    );
  }

  // ******************************************
  // END REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * Call the GraphQL mutation function and save the new state to the backend.
   * @param stationIds modified station ids
   * @param comment (optional) an optional comment for the acknowledgement
   */
  private readonly acknowledgeSohStatus = (stationNames: string[], comment?: string) => {
    // If station entries are not stale or already acknowledged then
    // call acknowledge mutation with station names
    if (
      isAcknowledgeEnabled(
        stationNames,
        this.props.sohStatus.stationAndStationGroupSoh.stationSoh,
        this.props.uiConfigurationQuery.uiAnalystConfiguration.sohStationStaleTimeMS
      )
    ) {
      this.props
        .acknowledgeSohStatus({
          variables: {
            stationNames,
            comment
          }
        })
        .catch(err => {
          UILogger.Instance().error(err);
        });
    } else {
      const toaster = new Toaster();
      toaster.toastWarn('Cannot acknowledge due to stale SOH data');
    }
  }

  /**
   * Check if SOH exceeds 1 minute display to UI criteria
   */
  private readonly checkSOHDisplayCriteria = (stationStatus: SohTypes.UiStationSoh[]): void => {
    // If UI Configuration has not loaded yet bail
    if (
      !this.props.uiConfigurationQuery ||
      !this.props.uiConfigurationQuery.uiAnalystConfiguration
    ) {
      return;
    }
    const messages: string[] = [];

    stationStatus.forEach(soh => {
      if (!this.sohCheckDisplayCriteriaMap.has(soh.stationName)) {
        // Set the entry, the first time we don't check
        this.sohCheckDisplayCriteriaMap.set(soh.stationName, soh.time);
      } else if (this.sohCheckDisplayCriteriaMap.get(soh.stationName) !== soh.time) {
        // Set the entry in the map so we don't check the entry again
        this.sohCheckDisplayCriteriaMap.set(soh.stationName, soh.time);

        // Get time and scale to milliseconds
        const deltaSecs = Date.now() / MILLISECONDS_IN_SECOND - soh.time;

        // add to avg
        this.avgSohLag += deltaSecs;
        this.sohStationsCount++;

        // Check min and max
        if (!this.minSohLag || deltaSecs < this.minSohLag) {
          this.minSohLag = deltaSecs;
        }

        if (!this.maxSohLag || deltaSecs > this.maxSohLag) {
          this.maxSohLag = deltaSecs;
        }
      }
    });

    // only log if there are messages to log
    if (messages && messages.length > 0) {
      UILogger.Instance().warn(...messages);
    }

    // Check if need to send statistic report
    if (Date.now() > this.lastSohReport + this.sohLagReportPeriod) {
      UILogger.Instance().info(
        `SOH lag statistics for last ${millisToStringWithMaxPrecision(this.sohLagReportPeriod)}. ` +
          `Max lag: ${millisToStringWithMaxPrecision(this.maxSohLag * MILLISECONDS_IN_SECOND)}, ` +
          `Min lag ${millisToStringWithMaxPrecision(this.minSohLag * MILLISECONDS_IN_SECOND)}, ` +
          `Avg lag ${millisToStringWithMaxPrecision(
            (this.avgSohLag / this.sohStationsCount) * MILLISECONDS_IN_SECOND
          )}`
      );

      // Reset statistics
      this.lastSohReport = Date.now();
      this.sohStationsCount = 0;
      this.avgSohLag = 0;
      this.minSohLag = undefined;
      this.maxSohLag = undefined;
    }
  }
}
