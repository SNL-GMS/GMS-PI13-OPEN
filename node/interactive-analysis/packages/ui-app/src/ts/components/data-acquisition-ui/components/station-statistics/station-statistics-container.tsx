import { compose } from '@gms/common-util';
import { WithNonIdealStates } from '@gms/ui-core-components';
import { AppState, DataAcquisitionWorkspaceOperations } from '@gms/ui-state';
import { withApollo } from 'react-apollo';
import * as ReactRedux from 'react-redux';
import { bindActionCreators } from 'redux';
import { CommonNonIdealStateDefs } from '~components/common-ui/components/non-ideal-states';
import { DataAcquisitionNonIdealStateDefs } from '~components/data-acquisition-ui/shared/non-ideal-states';
import { ReactApolloQueries } from '~components/react-apollo-components';
import { ReactApolloMutations } from '~data-acquisition-ui/react-apollo-components';
import { SohReduxProps } from '../../shared/types';
import { StationStatisticsComponent } from './station-statistics-component';
import { StationStatisticsProps } from './types';

/**
 * Mapping redux state to the properties of the component
 *
 * @param state App state, root level redux store
 */
const mapStateToProps = (state: AppState): Partial<SohReduxProps> => ({
  selectedStationIds: state.dataAcquisitionWorkspaceState.selectedStationIds,
  sohStatus: state.dataAcquisitionWorkspaceState.data.sohStatus
});

/**
 * Mapping methods (actions and operations) to dispatch one or more updates to the redux store
 *
 * @param dispatch the redux dispatch event alerting the store has changed
 */
const mapDispatchToProps = (dispatch): Partial<SohReduxProps> =>
  bindActionCreators(
    {
      setSelectedStationIds: DataAcquisitionWorkspaceOperations.setSelectedStationIds
    } as any,
    dispatch
  );

/**
 * Renders the station statistics display, or a non-ideal state from the provided list of
 * non ideal state definitions
 */
const StationStatisticsComponentOrNonIdealState = WithNonIdealStates<StationStatisticsProps>(
  [
    ...CommonNonIdealStateDefs.baseNonIdealStateDefinitions,
    ...DataAcquisitionNonIdealStateDefs.generalSohNonIdealStateDefinitions
  ],
  StationStatisticsComponent
);

/**
 * A new apollo component that's wrapping the Station Statistics component with
 * apollo graphQL queries and mutations.
 */
export const StationStatisticsContainer = compose(
  withApollo,
  ReactRedux.connect(mapStateToProps, mapDispatchToProps),
  ReactApolloMutations.graphqlAcknowledgeSohStatusMutation(),
  ReactApolloQueries.graphqlUIConfigurationQuery()
)(StationStatisticsComponentOrNonIdealState);
