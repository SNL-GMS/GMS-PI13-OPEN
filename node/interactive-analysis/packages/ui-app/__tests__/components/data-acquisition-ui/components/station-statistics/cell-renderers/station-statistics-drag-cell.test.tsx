import { uuid } from '@gms/common-util';
import { uniqueId } from 'lodash';
import React from 'react';
import {
  StationStatisticsDragCell,
  StationStatisticsDragCellProps
  // tslint:disable-next-line: max-line-length
} from '../../../../../../src/ts/components/data-acquisition-ui/components/station-statistics/cell-renderers/station-statistics-drag-cell';
import { StationStatisticsContext } from '../../../../../../src/ts/components/data-acquisition-ui/components/station-statistics/station-statistics-context';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');
uuid.asString = jest.fn().mockImplementation(uniqueId);

describe('Soh Drag Cell', () => {
  const dragCellProps: StationStatisticsDragCellProps = {
    stationId: 'AAA'
  };

  const wrapper = Enzyme.mount(
    <StationStatisticsContext.Provider
      value={{
        updateIntervalSecs: 1,
        quietTimerMs: 1,
        selectedStationIds: ['AAA'],
        setSelectedStationIds: jest.fn(),
        acknowledgeSohStatus: jest.fn(),
        sohStationStaleTimeMS: 30000
      }}
    >
      <StationStatisticsDragCell {...dragCellProps} />
    </StationStatisticsContext.Provider>
  );

  it('is defined', () => {
    expect(wrapper).toBeDefined();
  });

  it('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
