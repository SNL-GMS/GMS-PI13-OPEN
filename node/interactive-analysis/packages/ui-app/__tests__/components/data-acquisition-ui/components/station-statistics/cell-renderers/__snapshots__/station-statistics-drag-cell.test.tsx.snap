// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Soh Drag Cell Matches snapshot 1`] = `
<Component
  stationId="AAA"
>
  <DragCell
    getSelectedStationIds={[Function]}
    getSingleDragImage={[Function]}
    setSelectedStationIds={[Function]}
    stationId="AAA"
  >
    <DragInitiator
      getDragImage={[Function]}
      getDragPayload={[Function]}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <div
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      />
    </DragInitiator>
  </DragCell>
</Component>
`;
